realm-jni/src/io_realm_internal_table.cpp

```diff
@@ -144,6 +150,216 @@ JNIEXPORT void JNICALL Java_io_realm_internal_Table_nativeRenameColumn
+}
+
+JNIEXPORT void JNICALL Java_io_realm_internal_Table_nativeConvertColumnToNullable
+  (JNIEnv *env, jobject, jlong nativeTablePtr, jlong columnIndex)
+{
+    Table *table = TBL(nativeTablePtr);
+    if (!TBL_AND_COL_INDEX_VALID(env, table, columnIndex)) {
+        return;
+    }
+    if (table->has_shared_type()) {
+        ThrowException(env, UnsupportedOperation, "Not allowed to convert field in subtable.");
+        return;
+    }
+    try {
+        size_t column_index = S(columnIndex);
+        if (table->is_nullable(column_index)) {
+            return; // column is already nullable
+        }
+
+        std::string column_name = table->get_column_name(column_index);
+        DataType column_type = table->get_column_type(column_index);
+        if (column_type == type_Link ||
+            column_type == type_LinkList ||
+            column_type == type_Mixed ||
+            column_type == type_Table) {
+            ThrowException(env, IllegalArgument, "Wrong type - cannot be converted to nullable.");
+        }
+
+        std::string tmp_column_name;
+
+        size_t j = 0;
+        while (true) {
+            std::ostringstream ss;
+            ss << std::string("__TMP__") << j;
+            if (table->get_column_index(ss.str()) == realm::not_found) {
+                table->insert_column(column_index, column_type, ss.str(), true);
+                tmp_column_name = ss.str();
+                break;
+            }
+            j++;
+        }
+
+        for (size_t i = 0; i < table->size(); ++i) {
+            switch (column_type) {
+                case type_String:
+                    table->set_string(column_index, i, table->get_string(column_index + 1, i));
+                    break;
+                case type_Binary:
+                    table->set_binary(column_index, i, table->get_binary(column_index + 1, i));
+                    break;
+                case type_Int:
+                    table->set_int(column_index, i, table->get_int(column_index + 1, i));
+                    break;
+                case type_Bool:
+                    table->set_bool(column_index, i, table->get_bool(column_index + 1, i));
+                    break;
+                case type_DateTime:
+                    table->set_datetime(column_index, i, table->get_datetime(column_index + 1, i));
+                    break;
+                case type_Float:
+                    table->set_float(column_index, i, table->get_float(column_index + 1, i));
+                    break;
+                case type_Double:
+                    table->set_double(column_index, i, table->get_double(column_index + 1, i));
+                    break;
+                case type_Link:
+                case type_LinkList:
+                case type_Mixed:
+                case type_Table:
+                    // checked previously
+                    break;
+            }
+        }
+        if (table->has_search_index(column_index + 1)) {
```

Comment:
`column_index + 1` does not reference the old column always (if the tmp column was added at the end)
Ex: let's assume we have this table
`|__TMP__0|__TMP__1|`
if we want to migrate the second column the temp column will be add at position 2 
after inserting the tmp column we have 
`|__TMP__0|__TMP__1|__TMP__2`
`has_search_index(column_index + 1)` reference the new inserted column not the old one